cupy/core/_cub_reduction.pyx

```diff
@@ -0,0 +1,648 @@
+        reduce_type, params, arginfos, identity,
+        map_expr, reduce_expr, post_map_expr,
+        type_map, _preamble, options)
+
+
+cdef str _cub_path = _environment.get_cub_path()
+cdef str _cub_header = None
+
+
+cdef str _get_cub_header_include():
+    global _cub_header
+    if _cub_header is not None:
+        return _cub_header
+
+    assert _cub_path is not None
+    if _cub_path == '<bundle>':
+        _cub_header = '''
+#include <cupy/cub/cub/block/block_reduce.cuh>
+#include <cupy/cub/cub/block/block_load.cuh>
+'''
+    elif _cub_path == '<CUDA>':
+        _cub_header = '''
+#include <cub/block/block_reduce.cuh>
+#include <cub/block/block_load.cuh>
+'''
+    else:
+        _cub_header = '''
+#include \"${CUB}/cub/block/block_reduce.cuh\"
+#include \"${CUB}/cub/block/block_load.cuh\"
+'''
+        _cub_header = string.Template(_cub_header).substitute(CUB=_cub_path)
+    return _cub_header
+
+
+# make it cpdef'd for unit tests
+cdef inline tuple _can_use_cub_block_reduction(
+        list in_args, list out_args, tuple reduce_axis, tuple out_axis):
+    '''
+    If CUB BlockReduce can be used, this function returns a tuple of the needed
+    parameters, otherwise returns None.
+    '''
+    from cupy import core
+
+    cdef tuple axis_permutes_cub
+    cdef ndarray in_arr, out_arr
+    cdef Py_ssize_t contiguous_size = 1
+
+    # first check the flag settable at runtime from outside
+    if not core.cub_block_reduction_enabled:
+        return None
+
+    # detect whether CUB headers exists somewhere:
+    if _cub_path is None:
+        import warnings
+        warnings.warn('cupy.core.cub_block_reduction_enabled is set to True, '
+                      'but the CUB headers are not found, please set the '
+                      'environment variable CUPY_CUB_PATH to the CUB '
+                      'location.', RuntimeWarning)
+        return None
+
+    # rare event (mainly for conda-forge users): nvcc is not found!
+    if _environment.get_nvcc_path() is None:
+        return None
+
+    # we currently support only _SimpleReductionKernel
+    if len(in_args) != 1 or len(out_args) != 1:
+        return None
+
+    in_arr = in_args[0]
+    out_arr = out_args[0]
+
+    # To support generic reductions, note that some NumPy casting rules
+    # are not applicable in the C++ space (unless we tweak the type
+    # definitions). To circumvent this, we fall back to the old kernel.
+    # TODO(leofang): can we relax this?
+    if in_arr.dtype.kind != out_arr.dtype.kind:
+        # cannot cast complex to anything else
+        if in_arr.dtype.kind == 'c':
+            return None
+        # cannot cast float16 to complex
+        if in_arr.dtype.char == 'e' and out_arr.dtype.kind == 'c':
+            return None
+
+    # check reduction axes, if not contiguous then fall back to old kernel
+    reduce_axis = tuple(sorted(reduce_axis))
+    out_axis = tuple(sorted(out_axis))
+    if in_arr.flags.f_contiguous:
```

Comment:
This check can be performed earlier to improve cpu time if we have to fall back to cupy reduction.

also use `arr._f_contiguous`, `arr._c_contiguous` since accessing `flags` interacts with the python runtime.

Reply:
Fixed. PTAL.

Reply:
Thanks!